src/precice/impl/SolverInterfaceImpl.cpp

```diff
@@ -1172,6 +1179,247 @@ void SolverInterfaceImpl::writeScalarData(
+
+  // Get the data
+  WriteDataContext &context = _accessor->writeDataContext(dataID);
+  PRECICE_ASSERT(context.providedData() != nullptr);
+  mesh::Data &data = *context.providedData();
+
+  PRECICE_CHECK(data.hasGradient(), "Data \"{}\" has no gradient values available! Please turn on the gradient attribute in the configuration file.", data.getName())
+
+  PRECICE_CHECK(data.getDimensions() == 1,
+                "You cannot call writeBlockScalarGradientData on the vector data type \"{}\". Use writeBlockVectorGradientData or change the data type for \"{}\" to scalar.",
+                data.getName(), data.getName());
+
+  PRECICE_CHECK(data.getSpacialDimensions() == _dimensions,
+                "The spatial dimensions of the data object and of the mesh don't match. Please make sure to initialize {} space dimensions for the scalar data type \"{}\".",
+                _dimensions, context.getDataName());
+  PRECICE_VALIDATE_DATA(values, size * _dimensions);
+
+  // Get gradient data and check if initialized
+  auto &     gradientValuesInternal = data.gradientValues();
+  const auto vertexCount            = gradientValuesInternal.cols() / context.getDataDimensions();
+
+  for (int dim = 0; dim < _dimensions; dim++) {
+    for (int i = 0; i < size; i++) {
+
+      const auto valueIndex = valueIndices[i];
+      PRECICE_CHECK(0 <= valueIndex && valueIndex < vertexCount,
+                    "Cannot write gradient data \"{}\" to invalid Vertex ID ({}). Please make sure you only use the results from calls to setMeshVertex/Vertices().",
+                    context.getDataName(), valueIndex);
+
+      if (rowMajor) {
+        // Values are entered derived in spatial dimensions first
+        const int offset                        = dim * size;
+        gradientValuesInternal(dim, valueIndex) = values[offset + i];
+      } else {
+        // Values are entered components derived first
+        const int offset                        = i * _dimensions;
+        gradientValuesInternal(dim, valueIndex) = values[offset + dim];
+      }
+
+      //TODO: Add asserts
+    }
+  }
+}
+
+void SolverInterfaceImpl::writeVectorGradientData(
+    int           dataID,
+    int           valueIndex,
+    const double *value,
+    bool          rowMajor)
+{
+
+  PRECICE_TRACE(dataID, valueIndex);
+  PRECICE_CHECK(_state != State::Finalized, "writeVectorGradientData(...) cannot be called before finalize().")
+  PRECICE_REQUIRE_DATA_WRITE(dataID);
+
+  PRECICE_CHECK(value != nullptr, "writeVectorGradientData() was called with values == nullptr");
+
+  WriteDataContext &context = _accessor->writeDataContext(dataID);
+  PRECICE_ASSERT(context.providedData() != nullptr);
+  mesh::Data &data = *context.providedData();
+
+  // Check if Data object with ID dataID has been initialized with gradient data
+  PRECICE_CHECK(data.hasGradient(), "Data \"{}\" has no gradient values available! Please turn on the gradient attribute in the configuration file.", data.getName())
```

Comment:
```suggestion
  PRECICE_CHECK(data.hasGradient(), "Data \"{}\" has no gradient values available! Please turn on the gradient attribute in the configuration file.", data.getName())
```
```suggestion
  PRECICE_CHECK(data.hasGradient(), "Data \"{}\" has no gradient values available. Please turn on the gradient attribute in the configuration file.", data.getName())
```
Let's not use exclamation marks.